File: Data Structures & Algorithms/regular-expression-matching/submission-17.java
class Solution {

    int M,N;
    public boolean isMatch(String s, String p) {
        boolean[] dp;

        M = s.length();
        N=p.length();
        dp = new boolean[N+1];
        for(int i = M;i>=0;i--){
            boolean[] row = new boolean[N+1];
            if(i==M){
                row[N] = true;
            }
            for(int j = N-1;j>=0;j--){
                boolean isMatch = (i<M && (s.charAt(i)==p.charAt(j) || p.charAt(j) == '.'));
                if(j+1<N && p.charAt(j+1)=='*'){
                    row[j] = row[j] || (isMatch && dp[j]) || row[j+2];
                } else if(isMatch){
                    row[j] = row[j] || dp[j+1];
                } 
            }
            dp = row;
        }
        return dp[0];
    }

    // public boolean dfs(int i,int j, String s,String p){
    //     if(j==N){
    //         return i==M;
    //     }
    //     if(dp[i][j]!=null){
    //         return dp[i][j];
    //     }
    //     boolean isMatch = (i<M && (s.charAt(i)==p.charAt(j) || p.charAt(j) == '.'));
    //     if(j+1<N && p.charAt(j+1)=='*'){
    //         dp[i][j] = (isMatch && dfs(i+1,j,s,p)) || (dfs(i,j+2,s,p));
    //     }else if(isMatch){
    //         dp[i][j] = dfs(i+1,j+1,s,p);
    //     } else{
    //         dp[i][j] = false;
    //     }
    //     return dp[i][j];
    // }


    //             i

    // x   y   z   M

    // j
    // .   *   z
        
        
}

File: Data Structures & Algorithms/longest-increasing-subsequence/submission-17.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        N = len(nums)
        dp =[[0]*(N+1) for _ in range(N+1)]
        for i in range(N-1,-1,-1):
            for j in range(i-1,-2,-1):
                dp[i][j] = dp[i+1][j]
                if j == -1 or nums[i]>nums[j]:
                    dp[i][j] = max(dp[i][j],1+dp[i+1][i])
        return dp[0][-1]
        # def dfs(i,j):
        #     if (i,j) in dp:
        #         return dp[(i,j)]
        #     if i == N:
        #         return 0
        #     LIS = dfs(i+1,j)
        #     if j == -1 or nums[i]>nums[j]:
        #         LIS = max(LIS,1+dfs(i+1,i))
        #     dp[(i,j)] = LIS
        #     return dp[(i,j)]
        # return dfs(0,-1)
            
        